### PR TITLE
[roadmap] rework item on performance

### PR DESCRIPTION
**Pull request by gares**
This is an attempt of separating "trust" from "performance" in the main items.

What I want to say is that we optimize the system for the projects was regard as flagship. We have being doing so over the years relentlessly, it is nothing new ( @ppedrot , @SkySkimmer ), I'm just trying to put it into words and decorrelate it from trust.

Help from native speakers is more than welcome.

### CODE REVIEW

File: text/rocq-roadmap.md
Comment (Zimmi48) [inline text/rocq-roadmap.md:14]: Repetition of "We are committed to" on two subsequent lines (this one and the next that you did not change).

Furthermore, it should be "Performant" rather than "Performance" for consistency with the surrounding bullet points.

What about:

```suggestion
+ **Performant**: We optimize the system for our flagship projects in both industry and academia.
```

**Review by Zimmi48 — COMMENTED** [review]
These four bullet points are reflected in the structure of the document below and we should keep the link.

Do you think that it doesn't make sense to keep the two items in the same section? Then, what about moving performance in the same section as maintenance?

Otherwise, what about merging back these two lines, but keeping things more separated than they were, e.g.:

**Trustworthy and performant**: The Rocq Prover prioritizes self-verification, ensuring the highest level of trust in your formal proofs. At the same time, we optimize the system for our flagship projects in both industry and academia.

**Comment by Zimmi48** [comment]
And maybe remove the "our" in front of "flagship projects" to avoid carrying the meaning that we only care about projects that we own.

**Comment by gares** [comment]
I'm fine with your proposal.

> Trustworthy and performant: The Rocq Prover prioritizes self-verification, ensuring the highest level of trust in your formal proofs. At the same time, we optimize the system for our flagship projects in both industry and academia.

But even there you can see that the two concepts are not related (eg "At the same time"). So I'd prefer to move it to 4, i.e. make it part of usability. Point 3 has efficiency in the title, but looks like another meaning.

So my proposal is now:

**Usable**: We do our best to improve day-to-day usability for mathematicians, educators, and engineers alike. This includes improving performance, exploring AI-powered features and streamlined domain-specific tools.

### 4. Enhanced Usability

* **Performance**: optimize the system for flagship projects in both industry and academia.

**Review by Zimmi48 — APPROVED** [review]
There is already an item about performance in section 2 which should be removed or moved.

Besides this small issue, I'm happy to merge this change. I will do so if there are no more comments.

I think the "Enhanced usability" section is becoming too long and it probably means something, but it doesn't have to be addressed in this PR.

**Comment by gares** [comment]
I remove the line about performance from section 2.

W.r.t. section 2, I was wondering if we should mention the De Bruijn criteria (third party verification of proofs). But that is for another PR.

File: text/rocq-roadmap.md
Comment (jfehrle) [inline text/rocq-roadmap.md:51]: "optimize" isn't quite right--we want to improve performance where it makes a difference.  There's usually a tradeoff between performance and effort/maintainability.  The real optimization is that we should optimize to deliver the most value to users with the least effort.

Merely saying "improve performance" doesn't tell the reader much (yawn)--the same words could be used for almost any software project.  It needs more (brief) Coq-specific details to be interesting to read.  One example: something like making compilation of libraries significantly faster through increased parallelism, e.g. for the benefit of users such as Bedrock.  Extra points for giving a target, e.g. delivering a typical 10x speedup for large developments.

FWIW, "Usability" is primarily about making users more efficient in performing necessary tasks, e.g. adding features to refactor proof scripts would make users more efficient/productive.

Comment (jfehrle) [reply text/rocq-roadmap.md:51]: Also "flagship" is pretty vague.  I think we want to better support industrial-scale verification.  I don't know what the performance issues are for academia, whether they are much different from industry.  For academia, are there special performance needs for cutting edge research in type theory or mathematics?

Comment (gares) [reply text/rocq-roadmap.md:51]: I'll try to give examples:

Formalizing abstract mathematics, eg category theory, stresses some components of the system, like the universes-constraint checker. I don't think this math is relevant to the industry, at least not immediately. We have ongoing work to make this better/faster. UniMath will take advantage of this.

Verifying a compiler or an imperative program via Iris stresses other components. These use cases are relevant to the software industry.

To me flagship projects are projects "you should have heard of in your area of interest". CompCert, Iris if you are a PL person, MathComp, UniMath if you do math... 

In short I just wanted to say we focus the optimization efforts towards the flagship projects and that these encompasses both industrial verification projects and more academic ones like pure math. 

About usability, I know it is a bit on the edge, but for large verification efforts performance is key to usability (in the sense of being actually usable, not only pleasant to use).

Comment (jfehrle) [reply text/rocq-roadmap.md:51]:
```suggestion
* **Performance**: address performance limitations that impact users and in support
  of cutting-edge work such as:
  - large scale industrial software verification (faster build times)
  - formalization of category theory in UniComp (universe constraint checking)
  - MathComp [how would you describe this?]
  - verification of compilers and imperative programs (relying on the Iris package)
```

Maybe something like this?

Comment (gares) [reply text/rocq-roadmap.md:51]: Thanks, my counterproposal is:

**Performance**: address performance limitations that impact our users, especially in support
  of industrial software verification and large mathematical libraries.

Let's see what others think as well

Comment (silene) [reply text/rocq-roadmap.md:51]: I find both proposals to be oddly specific. So, here is my take on it: "address performance limitations that impacts projects that are widely used in industry and academia".

Comment (gares) [reply text/rocq-roadmap.md:51]: Maybe we could turn it into a more positive version, e.g. do not talk about limitation explicitly, but rather improve the perf so that things can go large(e) scale. But I'm not able to put that into (decent) words.

Comment (jfehrle) [reply text/rocq-roadmap.md:51]: > I find both proposals to be oddly specific.

The roadmap would much stronger with more detail.  We can be clear that these are things we would like to address in the near future but it is not guaranteed.

At some point, perhaps in a different document, we should identify which potential improvements to Rocq will provide the most value to users, their cost and rank them (without regard to whether anyone is already committed to implement them).

Comment (jfehrle) [reply text/rocq-roadmap.md:51]: > Maybe we could turn it into a more positive version, e.g. do not talk about limitation explicitly, but rather improve the perf so that things can go large(e) scale.

I suppose you could say "unmet performance needs of ...", but that's more of a euphemism, i.e., more vague and easier to misinterpret.  Every software package has performance limitations.  It's just a fact.  Saying so is not disparaging the software package.

**Comment by gares** [comment]
This seems to be the version @silene @theowinterhalter any myself like the most. @Zimmi48, @jfehrle are you ok with this last version?

File: text/rocq-roadmap.md
Comment (Zimmi48) [inline text/rocq-roadmap.md:51]: For consistency with the rest of the items, there should be target audiences. Given the content of the line, this should be:
```suggestion
* **Performance**: address performance limitations that impact projects that are widely used in industry and academia (C, D).
```

**Review by Zimmi48 — APPROVED** [review]
I'm quite satisfied by this proposal.

I want to emphasize that since this is a living document, we are aiming for incremental improvements. If everyone agrees this is an improvement, this PR should be merged.

If there are no more comments, I will merge at the end of the day.

**Comment by Zimmi48** [comment]
>The roadmap would much stronger with more detail. I don't think there's enough here so that users have much of an idea what is planned. We can be clear that the tasks it mentions are things we would like to address in the near future but it is not guaranteed we will get to them.
>
>At some point, perhaps in a different document, we should identify which potential improvements to Rocq will provide the most value to users, their cost and rank them (without regard to whether anyone is already committed to implement them).

For the record, this particular document aims to provide a high-level vision that should be reasonably stable in the next few years (except maybe, to remove the introduction which, like you noticed, is very specific to the renaming event).

This roadmap does not supersede the effort of #69 (which likely will find a different form, because a CEP has not been a good solution for a living document).

**Comment by Zimmi48** [comment]
> If there are no more comments, I will merge at the end of the day.

Actually, in this repo, we can self-merge PRs.

So @gares: please choose your preference among my two suggestions of audience above and then merge your PR.

**Comment by jfehrle** [comment]
> @gares [to me] are you ok with this last version?

> @Zimmi48 My take is that is better to start afresh and implement one (reasonable) behavior, document it, and recommend that for new code.

How about allowing me time to respond to @gares before merging?  Are you in a rush?  I was not available between 3 and 9 am PDT this morning when you posted your last comments.

**Comment by gares** [comment]
Sorry. Please open another pr with your comments and we will amend the text again.

**Comment by Zimmi48** [comment]
@jfehrle See what I wrote when approving the PR:

> I want to emphasize that since this is a living document, we are aiming for incremental improvements. If everyone agrees this is an improvement, this PR should be merged.

I should indeed have given you time to confirm whether you agree that the PR is an improvement.

But there is no point in delaying PRs that are considered improvements just because they could be improved further.

PRs are cheap, we can open new ones. If you have a vision of what the text should say, it's even easier if you are the author of the PR, rather than trying to convey your vision to someone else through review comments.

We are not adding new features to Coq here (where it makes sense to delay a PR until it is fully polished), but fixing a roadmap document that was already approved by the core team and which is already live on the Rocq website. Merging improvements quickly makes sense in this case.

**Comment by jfehrle** [comment]
I will survive.  However, I'm not so keen on submitting PRs for everything unlike the usual review process.  Often I want to make an observation or ask a question without having exact new wording.  That doesn't really fit with me submitting a PR.  Also it doesn't help that I have two fingers in a splint until maybe 1.5 weeks from now, so stuck doing hunt and peck for now.  Maybe I will comment later when I can type.  But also thinking the interaction takes considerable time, competing with other worthwhile tasks.  Suggesting wording improvements often takes significant time.

**Pull request merged**
